src/pgas.jl

```diff
@@ -0,0 +1,136 @@
+""" 
+    initialization(model::AbstractStateSpaceModel)
+
+Define the distribution of the initial state of the State Space Model
+"""
+function initialization end
+
+""" 
+    transition(model::AbstractStateSpaceModel, state, step)
+
+Define the transition density of the State Space Model
+Must return `nothing` if it consumed all the data
+"""
+function transition end
+
+"""
+    observation(model::AbstractStateSpaceModel, state, step)
+
+Return the log-likelihood of the observed measurement conditional on the current state of the model. 
+Must return `nothing` if it consumed all the data
+"""
+function observation end
+
+""" 
+    isdone(model::AbstractStateSpaceModel, step)
+
+Return `true` if model reached final state else `false`
+"""
+function isdone end
+
+"""
+    previous_state(trace::SSMTrace)
+
+Return `Xₜ₋₁` or `nothing` from `model`
+"""
+function previous_state(trace::SSMTrace)
+    return trace.f.X[step(trace) - 1]
+end
+
+"""
+    step(model::AbstractStateSpaceModel)
+
+Return current model step
+"""
+step(trace::SSMTrace) = trace.rng.count
```

Comment:
Would the `ref` particle store all previous `rng`'s? If so, does this `step` function work with the `ref` particle?